edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/InvokeMethodHandler.cs

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Core
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Net;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Device;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using static System.FormattableString;
+
+    public class InvokeMethodHandler : IInvokeMethodHandler
+    {
+        readonly ConcurrentDictionary<string, ConcurrentDictionary<DirectMethodRequest, TaskCompletionSource<DirectMethodResponse>>> clientMethodRequestQueue = new ConcurrentDictionary<string, ConcurrentDictionary<DirectMethodRequest, TaskCompletionSource<DirectMethodResponse>>>();
+        readonly IConnectionManager connectionManager;
+
+        public InvokeMethodHandler(IConnectionManager connectionManager)
+        {
+            this.connectionManager = Preconditions.CheckNotNull(connectionManager, nameof(connectionManager));
+        }
+
+        public Task<DirectMethodResponse> InvokeMethod(DirectMethodRequest methodRequest)
+        {
+            Preconditions.CheckNotNull(methodRequest, nameof(methodRequest));
+
+            Option<IDeviceProxy> deviceProxy = this.GetDeviceProxyWithSubscription(methodRequest.Id);
+            return deviceProxy.Map(
+                    d =>
+                    {
+                        Events.InvokingMethod(methodRequest);
+                        return d.InvokeMethodAsync(methodRequest);
+                    })
+                .GetOrElse(
+                    async () =>
+                    {
+                        var taskCompletion = new TaskCompletionSource<DirectMethodResponse>();
+                        ConcurrentDictionary<DirectMethodRequest, TaskCompletionSource<DirectMethodResponse>> clientQueue = this.GetClientQueue(methodRequest.Id);
+                        clientQueue.TryAdd(methodRequest, taskCompletion);
+                        Task completedTask = await Task.WhenAny(taskCompletion.Task, Task.Delay(methodRequest.ConnectTimeout));
+                        if (completedTask != taskCompletion.Task && clientQueue.TryRemove(methodRequest, out taskCompletion))
+                        {
+                            taskCompletion.TrySetResult(new DirectMethodResponse(new EdgeHubTimeoutException($"Client {methodRequest.Id} not found"), HttpStatusCode.NotFound));
+                        }
+
+                        return await taskCompletion.Task;
+                    });
+        }
+
+        public Task ProcessInvokeMethodSubscription(string id)
+        {
+            this.ProcessInvokeMethodSubscriptionInternal(id);
```

Comment:
Curious, why do we need a `ProcessInvokeMethodSubscriptionInternal` method?

Reply:
I suppose because you don't want to wait for the async operations inside that method to complete? Could we do a `Task.Run` or something instead so the intent is clearer?

Reply:
Yes, that is right. I can add a comment.